### PR TITLE
SE(2)-Equivariant Geometry Encoding: rotation-invariant relative features

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current model's OOD generalization failure — especially on tandem cases — may partly stem from implicit coordinate-frame dependence in the input features. DSDF uses absolute (x, y) coordinates; the TE coordinate frame (`--te_coord_frame`) helps but is incomplete. Adding SE(2)-equivariant geometric features ensures that rotating the entire tandem configuration produces consistently rotated outputs — a hard physical constraint currently enforced only softly through data augmentation.

For tandem, the **relative foil orientation (stagger angle)** is critical. Equivariant encoding makes this relationship explicit rather than requiring the model to learn it from data.

**This is a pure input feature addition — no architecture change, no new modules, zero risk of breaking torch.compile.**

**Literature support:**
- arXiv:2405.20287 (2024): "Flexible SE(2) graph neural networks with applications to PDE surrogates" (Bankestad, Mogren, Pirinen) — direct application to PDE surrogates, improved OOD generalization.
- ICML 2024 AI for Science Workshop: "SE(3)-Equivariant Diffusion Graph Nets" (Lino, Pfaff, Thuerey).
- NeurIPS 2024: "Space-Time Continuous PDE Forecasting using Equivariant Neural Fields" (Knigge et al., UvA).
- NSF Public Access: "Robust Emulator for Compressible Navier-Stokes using Equivariant Geometric Convolutions" (2024).

**Target metrics:** p_tan (primary — better inter-foil relative geometry encoding), p_oodc (secondary — rotation-invariant features help OOD generalization).

### 1. Add flag

```python
parser.add_argument('--se2_geometry', action='store_true',
    help='Add SE(2)-invariant relative angle features from each node to foil centroids')
```

### 2. Compute SE(2)-invariant features during preprocessing

Add the following features when `--se2_geometry` is active:

```python
def compute_se2_features(coords, boundary_id):
    """
    coords: [N, 2] — (x, y) of each mesh node
    boundary_id: [N] — 0=fore foil, 1=aft foil, -1=volume/boundary
    
    Returns: [N, 8] — SE(2)-invariant features
    """
    features = torch.zeros(coords.shape[0], 8, device=coords.device)
    
    # Compute foil centroids
    fore_mask = (boundary_id == 0)
    aft_mask = (boundary_id == 1)
    
    fore_centroid = coords[fore_mask].mean(dim=0) if fore_mask.any() else coords.mean(dim=0)
    aft_centroid = coords[aft_mask].mean(dim=0) if aft_mask.any() else coords.mean(dim=0)
    
    # Relative vectors from each node to each foil centroid
    dx_fore = coords[:, 0] - fore_centroid[0]
    dy_fore = coords[:, 1] - fore_centroid[1]
    dx_aft = coords[:, 0] - aft_centroid[0]
    dy_aft = coords[:, 1] - aft_centroid[1]
    
    # Distance (SE(2)-invariant)
    r_fore = (dx_fore**2 + dy_fore**2).sqrt() + 1e-8
    r_aft = (dx_aft**2 + dy_aft**2).sqrt() + 1e-8
    
    # Angle encoding (cos θ, sin θ) — equivariant under rotation
    features[:, 0] = dx_fore / r_fore  # cos θ_fore
    features[:, 1] = dy_fore / r_fore  # sin θ_fore
    features[:, 2] = r_fore             # distance to fore centroid (invariant)
    features[:, 3] = dx_aft / r_aft     # cos θ_aft
    features[:, 4] = dy_aft / r_aft     # sin θ_aft
    features[:, 5] = r_aft              # distance to aft centroid (invariant)
    
    # Inter-foil relative geometry (critical for tandem)
    foil_vec = aft_centroid - fore_centroid
    foil_dist = foil_vec.norm() + 1e-8
    features[:, 6] = foil_vec[0] / foil_dist  # cos of stagger angle
    features[:, 7] = foil_vec[1] / foil_dist  # sin of stagger angle
    
    # For single-foil samples (no aft foil), zero out aft-related features
    if not aft_mask.any():
        features[:, 3:8] = 0.0
    
    # Scale features for stability
    features[:, 2] = features[:, 2] * 0.1  # r_fore (raw distances can be large)
    features[:, 5] = features[:, 5] * 0.1  # r_aft
    
    return features
```

### 3. Integration

Add these 8 features to the input feature tensor in the same way other features (wake_deficit, vortex_panel, etc.) are concatenated. The model's `in_dim` should increase by 8.

**Important:** These features are computed from raw (x, y) coordinates and `boundary_id`, both of which are already available in the data loader. No new data files needed.

### 4. Handle single-foil samples

For single-foil samples, zero out the aft-related features (channels 3-7). Use the same pattern as `--cp_panel_tandem_only` or `--wake_deficit_feature` for masking single-foil samples.

### Run commands

```bash
# Seed 42
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent frieren --wandb_name "frieren/se2-geometry-s42" --wandb_group se2-geometry \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only \
  --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity \
  --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --se2_geometry

# Seed 73 (same flags, --seed 73, --wandb_name "frieren/se2-geometry-s73")
```

Note: **include `--surface_refine`** — this is a pure feature addition, the SRF head is preserved.

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.872** | < 11.872 |
| p_oodc | 7.459 | < 7.459 |
| **p_tan** | **26.319** | < 26.319 |
| **p_re** | **6.229** | < 6.229 |

W&B baseline: aycq1m8m (seed 42, best epoch 155), 9sk276v6 (seed 73, best epoch 156)

Reproduce:
```bash
cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output \
  --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Summary

**Neutral result** — SE(2)-invariant geometry features produce metrics within noise of baseline. p_oodc shows a tiny improvement (−0.8%), while p_in and p_tan regress slightly (+2.3%, +2.8%). The SE(2) features neither significantly help nor hurt.

### Metrics

| Metric | Seed 42 | Seed 73 | **2-seed avg** | Baseline | Δ |
|--------|---------|---------|---------------|----------|---|
| **p_in** | 12.5 | 11.8 | **12.15** | 11.872 | +0.28 (+2.3%) |
| **p_tan** | 27.1 | 27.0 | **27.05** | 26.319 | +0.73 (+2.8%) |
| **p_oodc** | 7.4 | 7.4 | **7.40** | 7.459 | −0.06 (−0.8%) ✓ |
| **p_re** | 6.2 | 6.3 | **6.25** | 6.229 | +0.02 (+0.3%) |

**In-distribution (seed avg, val_in_dist):** surf_Ux=2.64, surf_Uy=0.81, surf_p=12.15, vol_Ux=0.56, vol_Uy=0.20, vol_p=12.21

**Best epochs:** S42=153 (val/loss=0.5940), S73=153 (val/loss=0.5957) — both hit 180-min wall clock

### Train commands

```bash
# Seed 42
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent frieren --wandb_name "frieren/se2-geometry-s42" --wandb_group se2-geometry \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only \
  --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity \
  --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --se2_geometry

# Seed 73: same with --seed 73 --wandb_name "frieren/se2-geometry-s73"
```

### Peak memory

S42: 40.5 GB, S73: 40.0 GB (negligible overhead — just 8 extra float features per node)

### W&B run IDs

- Seed 42: `zbc14uib` (W&B) / `8o6g5fvm` (model dir)
- Seed 73: `18au6z1e`
- W&B group: `se2-geometry`

### What happened

The SE(2)-invariant relative geometry features (8 channels: cos/sin angles + scaled distances to fore/aft centroids + stagger direction) were cleanly implemented and integrated as additional input features. Training converged to the same quality as baseline — the features are not harmful, but don't provide meaningful lift.

**Why neutral?** The model already has substantial geometric context from existing features:
1. **TE coordinate frame** (`--te_coord_frame`): Already provides 6 features including dx/dy/r relative to TE points, which encode similar relative geometry.
2. **Fourier PE**: 32 channels of sin/cos positional encoding capture spatial relationships.
3. **DSDF (signed distance functions)**: 8 channels of distance/direction fields already encode foil proximity and shape.
4. **Wake deficit + angle**: 3 channels encode inter-foil relationship via trailing-edge frame.

The SE(2) features add redundant information. The cos(θ)/sin(θ) from each node to the centroid is essentially a normalized version of what TE coord features already capture, and the stagger direction (channels 6-7) is already implicit in the gap/stagger raw features (x[:, 22:24]).

**Note on SE(2) invariance**: The (cos θ, sin θ) features are actually SE(2)-**equivariant** (they co-rotate with the frame), not invariant. Only the distance features (r_fore, r_aft) are truly SE(2)-invariant. Since our training data is not augmented with random rotations, this equivariance property doesn't provide a useful inductive bias — the coordinate frame is fixed across all samples.

### Suggested follow-ups

1. **SE(2)-invariant features only** (distances, no angles): Test with just `r_fore, r_aft, foil_dist` (3 features instead of 8). This removes the redundant angle information and tests whether pure distance-based geometry helps.
2. **Replace TE coord frame**: Instead of adding SE(2) features alongside `--te_coord_frame`, try them as a replacement (without `--te_coord_frame`). The SE(2) features may be a cleaner encoding of similar information.
3. **Random rotation augmentation**: Pair SE(2)-equivariant features with SO(2) data augmentation (randomly rotate geometry+labels). This would create the setting where SE(2) equivariance actually provides an inductive bias.
4. **Relative inter-node features**: Instead of global centroid-based features, compute local SE(2)-invariant features between neighboring surface nodes (e.g., local curvature, inter-node distance patterns). This would provide new information not captured by existing features.
